docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,431 @@
+resolve the partition by deleting the node and the pods on the node.
+Alternatively, the actor may take steps to ensure the partitioned node
+cannot return to the cluster or access shared resources - this is known
+as **fencing** and is a well understood domain.
+
+This proposal covers the changes necessary to ensure:
+
+* Pet Sets can ensure **at most one** semantics for each individual pet
+* Other system components such as the node and namespace controller can
+  safely perform their responsibilities without violating that guarantee
+* An administrator or higher level controller can signal that a node
+  partition is permanent, allowing the Pet Set controller to proceed.
+* A fencing controller can take corrective action automatically to heal
+  partitions
+
+We will accomplish this by:
+
+* Clarifying which components are allowed to force delete pods (as opposed
+  to merely requesting termination)
+* Ensuring system components can observe partitioned pods and nodes
+  correctly
+* Defining how a fencing controller could safely interoperate with
+  partitioned nodes and pods to safely heal partitions
+* Describing how shared storage components without innate safety
+  guarantees can be safely shared on the cluster.
+
+
+### Current Guarantees for Pod lifecycle
+
+The existing pod model provides the following guarantees:
+
+* A pod is executed on exactly one node
+* A pod has the following lifecycle phases:
+  * Creation
+  * Scheduling
+  * Execution
+    * Init containers
+    * Application containers
+  * Termination
+  * Deletion
+* A pod can only move through its phases in order, and may not return
+  to an earlier phase.
+* A user may specify an interval on the pod called the **termination
+  grace period** that defines the minimum amount of time the pod will
+  have to complete the termination phase, and all components will honor
+  this interval.
+* Once a pod begins termination, its termination grace period can only
+  be shortened, not lengthened.
+
+Pod termination is divided into the following steps:
+
+* A component requests the termination of the pod by issuing a DELETE
+  to the pod resource with an optional **grace period**
+  * If no grace period is provided, the default from the pod is leveraged
+* When the kubelet observes the deletion, it starts a timer equal to the
+  grace period and performs the following actions:
+  * Executes the pre-stop hook, if specified, waiting up to **grace period**
+    seconds before continuing
+  * Sends the termination signal to the container runtime (SIGTERM or the
+    container image's STOPSIGNAL on Docker)
+  * Waits 2 seconds, or the remaining grace period, whichever is longer
+  * Sends the force termination signal to the container runtime (SIGKILL)
+* Once the kubelet observes the container is fully terminated, it issues
+  a status update to the REST API for the pod indicating termination, then
+  issues a DELETE with grace period = 0.
```

Comment:
The Kubelet doesn't need a finalizer, because the finalizer doesn't do anything.  Force delete has to trump finalizers (so app admins and fencers can resolve partitions), and since both fencers and app admins can trump finalizers, the finalizer is only symbolic and adds no value.